### PR TITLE
fix(sum): clamp -inf log-gradients in EM to prevent NaN weights

### PR DESCRIPTION
**Pull request by bewit**
During parameter optimization with Expectation-Maximization, the sum weights may underflow, leading to NaN weights. Clamping them to the minimum float value provided by PyTorch prevents that.

EDIT: After reevaluation, it is not the sum weights that underflow, but the likelihood gradients. In very deep PCs (e.g. a chain-structured PC of length 1024), the gradients may vanish. The module_lls.grad being 0 leads to their log becoming -inf, which in turn is transformed to NaN in the following log_expectations.log_softmax(). This NaN is propagated to the sum weights in the EM update. The clamp prevents this error, but does not mitigate the vanishing gradients. We might replace the minimum float value with an epsilon parameter later.

### CODE REVIEW

**Comment by braun-steven** [comment]
Are you sure this is an underflow? I would be surprised since these are _log_-gradients already.

**Comment by bewit** [comment]
I encountered this issue when trying to implement the Hidden Markov Model structured PCs that are used by recent works from the StarAI lab. They are extremely "deep", as at each depth, we are splitting away only one random variable. For 1024 random variables, we have 1024 layers, for which the gradients might vanish. 
I would say it is an underflow. But we could also look for alternatives, e.g. to apply this fix only in the HMM implementation in the zoo or recommend AnEMone to train them.

EDIT: Let me check again if it is an underflow or other issue.

**Comment by bewit** [comment]
I have checked again and it is not an underflow in the sum weights per se, but the sum weights becoming NaN is the consequence of an underflow/vanishing gradient problem in module_lls.grad. The gradients being '0' leads to log_grads becoming '-inf', which then become 'NaN' in log_expectations.log_softmax().
This is only a problem in very deep networks like Hidden Markov Model-structured PCs.

@braun-steven Do you have any recommendations how to handle this? If not, I would keep this fix in place and edit the PR message.

**Comment by braun-steven** [comment]
Clamping to some min value should be fine I guess. Not sure if some some epsilon or finfo of float min is better here. If finfo works for you I'd keep it.
